ddl/partition.go

```diff
@@ -1115,22 +1115,45 @@ func setPartitionPlacementFromOptions(partition *model.PartitionDefinition, opti
 	return nil
 }
 
-func buildHashPartitionDefinitions(_ sessionctx.Context, defs []*ast.PartitionDefinition, tbInfo *model.TableInfo) ([]model.PartitionDefinition, error) {
+func isNonDefaultPartitionOptionsUsed(defs []model.PartitionDefinition) bool {
+	for i := range defs {
+		orgDef := defs[i]
+		if orgDef.Name.O != fmt.Sprintf("p%d", i) {
+			return true
+		}
+		if len(orgDef.Comment) > 0 {
+			return true
+		}
+		if orgDef.PlacementPolicyRef != nil {
+			return true
+		}
+	}
+	return false
+}
+
+func buildHashPartitionDefinitions(_ sessionctx.Context, defs []*ast.PartitionDefinition, tbInfo *model.TableInfo, numParts uint64) ([]model.PartitionDefinition, error) {
 	if err := checkAddPartitionTooManyPartitions(tbInfo.Partition.Num); err != nil {
 		return nil, err
 	}
 
-	definitions := make([]model.PartitionDefinition, tbInfo.Partition.Num)
-	for i := 0; i < len(definitions); i++ {
-		if len(defs) == 0 {
-			definitions[i].Name = model.NewCIStr(fmt.Sprintf("p%v", i))
-		} else {
-			def := defs[i]
+	definitions := make([]model.PartitionDefinition, numParts)
+	// So start by re-using the existing ones
+	// For COALESCE, it will copy the remaining ones
+	// For ADD it will copy the existing ones
+	copy(definitions, tbInfo.Partition.Definitions)
+	offset := uint64(len(tbInfo.Partition.Definitions))
+
+	// ADD PARTITION needs to add the new partitions
+	for i := offset; i < numParts; i++ {
+		if (i - offset) < uint64(len(defs)) {
```

Comment:
Is there any case which will cover the `if (i - offset) < uint64(len(defs))...` code block and `else` code block at the same time. If no, why dont' separate the two from loop.

Reply:
There should not be, since number of partitions `partitions N` and partition definitions `(partition p1 ..., ...)` are mutual exclusive.

But I rewrote it slightly to avoid copying all information, and only copying name, comment and placement rules.